src/redisearch_rs/slots_tracker/src/slot_set.rs

```diff
@@ -0,0 +1,1040 @@
+ * GNU Affero General Public License v3 (AGPLv3).
+*/
+
+//! Internal implementation of the SlotSet data structure.
+//!
+//! This module contains the private implementation of slot range tracking.
+//! It is not exposed outside of the slots_tracker crate.
+
+use crate::SlotRange;
+
+/// Enum describing the relationship between a set and a query.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub(crate) enum CoverageRelation {
+    Equals,  // Set == query (exact match)
+    Covers,  // Set ⊇ query (covers but has extra)
+    NoMatch, // Set ⊉ query (doesn't cover all)
+}
+
+// ============================================================================
+// Debug assertion helpers
+// ============================================================================
+
+/// Validates that all ranges are valid.
+#[inline]
+fn debug_assert_valid_ranges(ranges: &[SlotRange]) {
+    debug_assert!(ranges.iter().all(|r| r.start <= r.end && r.end <= 16383));
+}
+
+/// Validates that ranges are sorted and normalized (no overlaps or adjacent ranges).
+#[inline]
+fn debug_assert_normalized_ranges(ranges: &[SlotRange]) {
+    debug_assert!(ranges.windows(2).all(|w| w[0].end + 1 < w[1].start));
+}
+
+/// Validates that ranges are valid, sorted, and normalized.
+#[inline]
+fn debug_assert_valid_normalized_input(ranges: &[SlotRange]) {
+    debug_assert_valid_ranges(ranges);
+    debug_assert_normalized_ranges(ranges);
+}
+
+/// A collection of slot ranges with set operation capabilities.
+///
+/// This is an internal type used only within this crate to manage slot sets.
+/// Each range is inclusive [start, end].
+///
+/// **Invariant**: The ranges vector is always kept sorted and normalized:
+/// - Sorted by start slot in ascending order
+/// - No overlapping ranges
+/// - No adjacent ranges (they are merged)
+/// - All ranges are valid (start <= end, values in [0, 16383])
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+pub(crate) struct SlotSet {
+    /// Vector of slot ranges (start, end), kept sorted and normalized.
+    ranges: Vec<SlotRange>,
+}
+
+impl SlotSet {
+    /// Creates a new empty `SlotSet`.
+    pub(crate) const fn new() -> Self {
+        Self { ranges: Vec::new() }
+    }
+
+    // ========================================================================
+    // Public API methods required for the C FFI:
+    // ========================================================================
+
+    /// Replaces the entire contents of this SlotSet with the given ranges (hard reset).
+    ///
+    /// **Assumes input is already sorted and normalized** (no overlaps, no adjacent ranges).
+    /// In debug builds, validates this assumption.
+    /// Used only by `slots_tracker_set_local_slots`.
+    pub(crate) fn set_from_ranges(&mut self, ranges: &[SlotRange]) {
+        debug_assert_valid_normalized_input(ranges);
+
+        // Input is already normalized, just replace our vector
+        self.ranges = ranges.to_vec();
+    }
```

Comment:
```suggestion
    pub(crate) fn from_ranges(ranges: &[SlotRange]) -> Self {
        debug_assert_valid_normalized_input(ranges);

        // Input is already normalized, just replace our vector
        Self { ranges:ranges.to_vec() }
    }
```

This overrides everything so can just be a constructor instead